File: implementation/01 Fundamentals/Concurreny Basics/Thread-safe Counter/go/counter_test.go
package counter

import (
	"sync"
	"testing"
)

// Test constants — high enough to make races statistically certain to manifest.
const (
	threadCount   = 10
	opsPerThread  = 100_000
	expectedTotal = int64(threadCount * opsPerThread)
)

// Run these tests with: go test -race ./...
// The -race flag enables Go's built-in race detector.
// UnsafeCounter will trigger race detector warnings immediately.
// MutexCounter and AtomicCounter should produce zero race warnings.

// =============================================================================
// Shared test helpers
// =============================================================================

// runConcurrent launches n goroutines that all start simultaneously via WaitGroup
// and each execute fn opsPerThread times. Blocks until all goroutines complete.
//
// sync.WaitGroup pattern:
//   wg.Add(n)        — register n goroutines to wait for
//   go func() {
//     defer wg.Done() — signal completion
//     ...
//   }()
//   wg.Wait()        — block until Done() called n times
func runConcurrent(n int, fn func()) {
	// ready channel acts as a gate — all goroutines wait until closed simultaneously.
	// This maximizes contention — worst-case scenario for the synchronization mechanism.
	ready := make(chan struct{})
	var wg sync.WaitGroup
	wg.Add(n)

	for i := 0; i < n; i++ {
		go func() {
			defer wg.Done()
			<-ready // block until gate opens
			for j := 0; j < opsPerThread; j++ {
				fn()
			}
		}()
	}

	close(ready) // release all goroutines simultaneously
	wg.Wait()
}

// =============================================================================
// Functional correctness — single goroutine
// =============================================================================

func TestMutexCounter_Increment(t *testing.T) {
	c := &MutexCounter{}
	c.Increment()
	c.Increment()
	c.Increment()
	if got := c.Get(); got != 3 {
		t.Errorf("expected 3, got %d", got)
	}
}

func TestAtomicCounter_Increment(t *testing.T) {
	c := &AtomicCounter{}
	c.Increment()
	c.Increment()
	c.Increment()
	if got := c.Get(); got != 3 {
		t.Errorf("expected 3, got %d", got)
	}
}

func TestCounter_Decrement_BelowZero(t *testing.T) {
	counters := []Counter{&MutexCounter{}, &AtomicCounter{}}
	for _, c := range counters {
		c.Decrement()
		if got := c.Get(); got != -1 {
			t.Errorf("expected -1, got %d", got)
		}
	}
}

func TestCounter_IncrementBy_ValidDelta(t *testing.T) {
	counters := []Counter{&MutexCounter{}, &AtomicCounter{}}
	for _, c := range counters {
		if err := c.IncrementBy(50); err != nil {
			t.Errorf("unexpected error: %v", err)
		}
		if got := c.Get(); got != 50 {
			t.Errorf("expected 50, got %d", got)
		}
	}
}

func TestCounter_IncrementBy_InvalidDelta(t *testing.T) {
	counters := []Counter{&MutexCounter{}, &AtomicCounter{}}
	for _, c := range counters {
		if err := c.IncrementBy(0); err == nil {
			t.Error("expected error for delta=0, got nil")
		}
		if err := c.IncrementBy(-5); err == nil {
			t.Error("expected error for delta=-5, got nil")
		}
	}
}

func TestCounter_Reset(t *testing.T) {
	counters := []Counter{&MutexCounter{}, &AtomicCounter{}}
	for _, c := range counters {
		_ = c.IncrementBy(1000)
		c.Reset()
		if got := c.Get(); got != 0 {
			t.Errorf("expected 0 after reset, got %d", got)
		}
	}
}

// =============================================================================
// Concurrency correctness — thread-safe implementations must lose zero increments
// =============================================================================

func TestMutexCounter_HighConcurrency_NoLostUpdates(t *testing.T) {
	c := &MutexCounter{}
	runConcurrent(threadCount, c.Increment)
	if got := c.Get(); got != expectedTotal {
		t.Errorf("lost increments: expected %d, got %d (lost %d)",
			expectedTotal, got, expectedTotal-got)
	}
}

func TestAtomicCounter_HighConcurrency_NoLostUpdates(t *testing.T) {
	c := &AtomicCounter{}
	runConcurrent(threadCount, c.Increment)
	if got := c.Get(); got != expectedTotal {
		t.Errorf("lost increments: expected %d, got %d (lost %d)",
			expectedTotal, got, expectedTotal-got)
	}
}

func TestMutexCounter_MixedOperations_CorrectNetValue(t *testing.T) {
	// Half goroutines increment, half decrement. Net must be zero.
	c := &MutexCounter{}
	ready := make(chan struct{})
	var wg sync.WaitGroup
	wg.Add(threadCount)

	for i := 0; i < threadCount; i++ {
		increment := i%2 == 0
		go func(inc bool) {
			defer wg.Done()
			<-ready
			for j := 0; j < opsPerThread; j++ {
				if inc {
					c.Increment()
				} else {
					c.Decrement()
				}
			}
		}(increment)
	}

	close(ready)
	wg.Wait()

	if got := c.Get(); got != 0 {
		t.Errorf("mixed ops should net to 0, got %d", got)
	}
}

func TestAtomicCounter_MixedOperations_CorrectNetValue(t *testing.T) {
	c := &AtomicCounter{}
	ready := make(chan struct{})
	var wg sync.WaitGroup
	wg.Add(threadCount)

	for i := 0; i < threadCount; i++ {
		increment := i%2 == 0
		go func(inc bool) {
			defer wg.Done()
			<-ready
			for j := 0; j < opsPerThread; j++ {
				if inc {
					c.Increment()
				} else {
					c.Decrement()
				}
			}
		}(increment)
	}

	close(ready)
	wg.Wait()

	if got := c.Get(); got != 0 {
		t.Errorf("mixed ops should net to 0, got %d", got)
	}
}

// =============================================================================
// AtomicCounter-specific: GetAndReset
// =============================================================================

func TestAtomicCounter_GetAndReset_ReturnsSnapshotAndZeros(t *testing.T) {
	c := &AtomicCounter{}
	_ = c.IncrementBy(500)

	snapshot := c.GetAndReset()

	if snapshot != 500 {
		t.Errorf("expected snapshot=500, got %d", snapshot)
	}
	if got := c.Get(); got != 0 {
		t.Errorf("expected 0 after GetAndReset, got %d", got)
	}
}

func TestAtomicCounter_GetAndReset_NoIncrementLost(t *testing.T) {
	// Concurrent increments + GetAndReset: snapshot + remaining must equal total.
	c := &AtomicCounter{}
	incrementGoroutines := 4
	opsEach := 10_000
	totalOps := int64(incrementGoroutines * opsEach)

	ready := make(chan struct{})
	var wg sync.WaitGroup
	wg.Add(incrementGoroutines)

	for i := 0; i < incrementGoroutines; i++ {
		go func() {
			defer wg.Done()
			<-ready
			for j := 0; j < opsEach; j++ {
				c.Increment()
			}
		}()
	}

	close(ready)
	wg.Wait()

	snapshot := c.GetAndReset()
	remaining := c.Get()

	// Any increment that raced with GetAndReset lands in remaining.
	if snapshot+remaining != totalOps {
		t.Errorf("lost increments across GetAndReset: snapshot=%d + remaining=%d = %d, want %d",
			snapshot, remaining, snapshot+remaining, totalOps)
	}
}

// =============================================================================
// Benchmarks — run with: go test -bench=. -benchmem
// =============================================================================

func BenchmarkMutexCounter_Increment(b *testing.B) {
	c := &MutexCounter{}
	b.RunParallel(func(pb *testing.PB) {
		for pb.Next() {
			c.Increment()
		}
	})
}

func BenchmarkAtomicCounter_Increment(b *testing.B) {
	c := &AtomicCounter{}
	b.RunParallel(func(pb *testing.PB) {
		for pb.Next() {
			c.Increment()
		}
	})
}

// BenchmarkAtomicCounter_GetAndReset measures the CAS retry loop cost.
func BenchmarkAtomicCounter_GetAndReset(b *testing.B) {
	c := &AtomicCounter{}
	_ = c.IncrementBy(1_000_000)
	b.RunParallel(func(pb *testing.PB) {
		for pb.Next() {
			c.Increment()
			if c.Get() > 10_000 {
				c.GetAndReset()
			}
		}
	})
}

File: implementation/01 Fundamentals/Concurreny Basics/Thread-safe Counter/java/CounterTest.java
// CounterTest.java

import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.MethodSource;

import java.util.concurrent.CountDownLatch;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.TimeUnit;
import java.util.stream.Stream;

import static org.junit.jupiter.api.Assertions.*;

/**
 * Test strategy:
 *
 * 1. Functional correctness — single-threaded, verifies basic operations.
 * 2. Concurrency correctness — multi-threaded, verifies no lost updates.
 * 3. Race condition demonstration — proves UnsafeCounter loses increments.
 * 4. Edge cases — invalid inputs, boundary behavior.
 *
 * Key testing challenge: concurrency bugs are non-deterministic.
 * We use high thread counts and high operation counts to make races
 * statistically certain to manifest, not just occasionally possible.
 *
 * CountDownLatch pattern:
 *   All threads wait at a gate (startLatch) until released simultaneously.
 *   This maximizes thread contention — the worst-case scenario.
 *   Without it, threads start staggered and rarely actually contend.
 */
class CounterTest {

    private static final int THREAD_COUNT = 10;
    private static final int OPS_PER_THREAD = 100_000;
    private static final long EXPECTED_TOTAL = (long) THREAD_COUNT * OPS_PER_THREAD;

    // Parameterized: run same concurrency tests against all thread-safe implementations.
    static Stream<Counter> threadSafeCounters() {
        return Stream.of(new MutexCounter(), new AtomicCounter());
    }

    // --- Functional correctness (single-threaded) ---

    @ParameterizedTest
    @MethodSource("threadSafeCounters")
    void increment_singleThread_correctValue(Counter counter) {
        counter.increment();
        counter.increment();
        counter.increment();
        assertEquals(3, counter.get());
    }

    @ParameterizedTest
    @MethodSource("threadSafeCounters")
    void decrement_belowZero_allowed(Counter counter) {
        counter.decrement();
        assertEquals(-1, counter.get(), "Counters should support negative values");
    }

    @ParameterizedTest
    @MethodSource("threadSafeCounters")
    void incrementBy_validDelta_correctValue(Counter counter) {
        counter.incrementBy(50);
        assertEquals(50, counter.get());
    }

    @ParameterizedTest
    @MethodSource("threadSafeCounters")
    void incrementBy_zeroDelta_throwsException(Counter counter) {
        assertThrows(IllegalArgumentException.class, () -> counter.incrementBy(0));
    }

    @ParameterizedTest
    @MethodSource("threadSafeCounters")
    void incrementBy_negativeDelta_throwsException(Counter counter) {
        assertThrows(IllegalArgumentException.class, () -> counter.incrementBy(-5));
    }

    @ParameterizedTest
    @MethodSource("threadSafeCounters")
    void reset_afterIncrements_returnsZero(Counter counter) {
        counter.incrementBy(1000);
        counter.reset();
        assertEquals(0, counter.get());
    }

    // --- Concurrency correctness ---

    @ParameterizedTest
    @MethodSource("threadSafeCounters")
    void increment_highConcurrency_noLostUpdates(Counter counter)
            throws InterruptedException {

        ExecutorService executor = Executors.newFixedThreadPool(THREAD_COUNT);
        CountDownLatch startLatch = new CountDownLatch(1);  // gate — holds all threads
        CountDownLatch doneLatch = new CountDownLatch(THREAD_COUNT);

        for (int i = 0; i < THREAD_COUNT; i++) {
            executor.submit(() -> {
                try {
                    startLatch.await();  // all threads wait here until released
                    for (int j = 0; j < OPS_PER_THREAD; j++) {
                        counter.increment();
                    }
                } catch (InterruptedException e) {
                    Thread.currentThread().interrupt();
                } finally {
                    doneLatch.countDown();
                }
            });
        }

        startLatch.countDown();  // release all threads simultaneously — maximum contention
        assertTrue(doneLatch.await(30, TimeUnit.SECONDS), "Threads did not complete in time");
        executor.shutdown();

        assertEquals(EXPECTED_TOTAL, counter.get(),
            "Expected " + EXPECTED_TOTAL + " but got " + counter.get() +
            " — lost " + (EXPECTED_TOTAL - counter.get()) + " increments");
    }

    @ParameterizedTest
    @MethodSource("threadSafeCounters")
    void mixedIncrementDecrement_highConcurrency_correctNetValue(Counter counter)
            throws InterruptedException {

        // Half threads increment, half decrement — net result must be zero.
        ExecutorService executor = Executors.newFixedThreadPool(THREAD_COUNT);
        CountDownLatch startLatch = new CountDownLatch(1);
        CountDownLatch doneLatch = new CountDownLatch(THREAD_COUNT);

        for (int i = 0; i < THREAD_COUNT; i++) {
            final boolean shouldIncrement = (i % 2 == 0);
            executor.submit(() -> {
                try {
                    startLatch.await();
                    for (int j = 0; j < OPS_PER_THREAD; j++) {
                        if (shouldIncrement) counter.increment();
                        else counter.decrement();
                    }
                } catch (InterruptedException e) {
                    Thread.currentThread().interrupt();
                } finally {
                    doneLatch.countDown();
                }
            });
        }

        startLatch.countDown();
        assertTrue(doneLatch.await(30, TimeUnit.SECONDS), "Threads did not complete in time");
        executor.shutdown();

        assertEquals(0, counter.get(),
            "Mixed increment/decrement should net to zero, got: " + counter.get());
    }

    // --- Race condition demonstration ---

    @Test
    void unsafeCounter_highConcurrency_losesIncrements() throws InterruptedException {
        // This test EXPECTS the unsafe counter to produce a wrong result.
        // It is not a bug in the test — it is proof that UnsafeCounter is broken.
        // If this test fails (unsafe counter returns correct value), the JVM
        // happened to serialize operations by luck — re-run to see the race.
        UnsafeCounter unsafe = new UnsafeCounter();
        ExecutorService executor = Executors.newFixedThreadPool(THREAD_COUNT);
        CountDownLatch startLatch = new CountDownLatch(1);
        CountDownLatch doneLatch = new CountDownLatch(THREAD_COUNT);

        for (int i = 0; i < THREAD_COUNT; i++) {
            executor.submit(() -> {
                try {
                    startLatch.await();
                    for (int j = 0; j < OPS_PER_THREAD; j++) {
                        unsafe.increment();
                    }
                } catch (InterruptedException e) {
                    Thread.currentThread().interrupt();
                } finally {
                    doneLatch.countDown();
                }
            });
        }

        startLatch.countDown();
        assertTrue(doneLatch.await(30, TimeUnit.SECONDS));
        executor.shutdown();

        long actual = unsafe.get();
        System.out.printf(
            "UnsafeCounter: expected=%d, actual=%d, lost=%d (%.1f%% loss)%n",
            EXPECTED_TOTAL, actual, EXPECTED_TOTAL - actual,
            (double)(EXPECTED_TOTAL - actual) / EXPECTED_TOTAL * 100
        );

        // We assert it is WRONG. If this assertion fails, the race didn't manifest —
        // increase THREAD_COUNT or OPS_PER_THREAD.
        assertNotEquals(EXPECTED_TOTAL, actual,
            "UnsafeCounter should have lost increments under concurrency");
    }

    // --- AtomicCounter-specific behavior ---

    @Test
    void atomicCounter_getAndReset_returnsSnapshotAndZeros() throws InterruptedException {
        AtomicCounter counter = new AtomicCounter();
        counter.incrementBy(500);

        long snapshot = counter.getAndReset();

        assertEquals(500, snapshot, "getAndReset should return value before reset");
        assertEquals(0, counter.get(), "counter should be zero after getAndReset");
    }

    @Test
    void atomicCounter_getAndReset_atomicity_noIncrementLost() throws InterruptedException {
        // Verifies that getAndReset doesn't lose concurrent increments.
        // After getAndReset completes, all increments that happened before
        // the reset are captured in the snapshot. Increments after are in the counter.
        AtomicCounter counter = new AtomicCounter();
        int incrementThreads = 4;
        ExecutorService executor = Executors.newFixedThreadPool(incrementThreads + 1);
        CountDownLatch startLatch = new CountDownLatch(1);
        CountDownLatch doneLatch = new CountDownLatch(incrementThreads);

        for (int i = 0; i < incrementThreads; i++) {
            executor.submit(() -> {
                try {
                    startLatch.await();
                    for (int j = 0; j < 10_000; j++) counter.increment();
                } catch (InterruptedException e) {
                    Thread.currentThread().interrupt();
                } finally {
                    doneLatch.countDown();
                }
            });
        }

        startLatch.countDown();
        doneLatch.await(10, TimeUnit.SECONDS);
        executor.shutdown();

        long snapshot = counter.getAndReset();
        long remaining = counter.get();

        // snapshot + remaining must equal total increments.
        // remaining captures any increments that occurred after getAndReset().
        assertEquals(incrementThreads * 10_000L, snapshot + remaining,
            "No increments should be lost across getAndReset boundary");
    }
}